[SCM] Using subfolder with a repository together with the conanfile

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Working with Conan 1.4.4 (also tested with 1.4.3) on OSX, we expect to rely on the (experimental) [SCM attribute of `ConanFile`](https://docs.conan.io/en/latest/creating_packages/package_repo.html#capturing-the-remote-and-commit-from-git-scm-experimental) to implement CI of our repositories (#3003 for context)

Here is a minimal conanfile:
```
from conans import ConanFile, CMake, tools

class ScmtestConan(ConanFile):
    name = "scm-test"
    version = "0.0.0"
    description = "Try the SCM conanfile attribute"
    settings = "os", "compiler", "build_type", "arch"

    generators = "cmake", "cmake_paths"

    scm = {
        "type": "git",
        "subfolder": "cloned_repo",
        "url": "auto",
        "revision": "auto"
    }

    def build(self):
        cmake = CMake(self)
        cmake.configure(source_folder="cloned_repo")
        cmake.build()
        cmake.install()

    def package(self):
        # Done by the install step in build()
        pass

    def package_info(self):
        self.cpp_info.libs = tools.collect_libs(self)
```

Now, if we export and try to install the recipe with `conan export . com/develop && conan install scm-test/0.0.0@com/develop`, the build fails.

### subfolder problem

After inspecting the created directories under `~/.conan`, it turn out that, even though a `cloned_repo` is created in the `source` folder, the repo is actually cloned directly into the `source` folder, and `cloned_repo` remains empty.

### same computer optimization problem

Now, if we were to move the conanfile.py from the root of the repo to a `conan/` subfolder, as we intend to, nothing is cloned at all, because it seems (following the TIP) that conan wants to use the local folder when installing on the same computer that did the export, and it uses the folder containing the conanfile, not necessarily the folder at the root of the repo.

Here is what is displayed in the terminal:
(please remember that I am on OSX - 10.12, in case it is os specific)

### Export

> $ conan export . com/develop
> WARN: Conanfile doesn't have 'url'.
> It is recommended to add it as attribute
> WARN: Conanfile doesn't have 'license'.
> It is recommended to add it as attribute
> scm-test/0.0.0@com/develop: Repo origin deduced by 'auto': git@github.com:Adnn/issue_conan-3069.git
> scm-test/0.0.0@com/develop: Revision deduced by 'auto': 2ee94f8181e70af0e6e0cb8d42a4e6a360e8fa68
> scm-test/0.0.0@com/develop: A new conanfile.py version was exported
> scm-test/0.0.0@com/develop: Folder: /Users/adn/.conan/data/scm-test/0.0.0/com/develop/export
> 
> $ git rev-parse --short HEAD
> 2ee94f8

### Install

$ conan install scm-test/0.0.0@com/develop --build
scm-test/0.0.0@com/develop: Installing package
Requirements
    scm-test/0.0.0@com/develop from local cache
Packages
    scm-test/0.0.0@com/develop:7fe67dff831b24bc4a8b5db678a51f1be5e44e7c

scm-test/0.0.0@com/develop: WARN: Forced build from source
scm-test/0.0.0@com/develop: Building your package in /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c
scm-test/0.0.0@com/develop: Configuring sources in /Users/adn/.conan/data/scm-test/0.0.0/com/develop/source
scm-test/0.0.0@com/develop: Getting sources from folder: /Users/adn/projects/linux_todel/issue_conan-3069
scm-test/0.0.0@com/develop: Copying sources to build folder
scm-test/0.0.0@com/develop: Generator cmake created conanbuildinfo.cmake
scm-test/0.0.0@com/develop: Generator cmake_paths created conan_paths.cmake
scm-test/0.0.0@com/develop: Calling build()
CMake Error: The source directory "/Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/cloned_repo" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
scm-test/0.0.0@com/develop:
scm-test/0.0.0@com/develop: ERROR: Package '7fe67dff831b24bc4a8b5db678a51f1be5e44e7c' build failed
scm-test/0.0.0@com/develop: WARN: Build folder /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c
ERROR: scm-test/0.0.0@com/develop: Error in build() method, line 18
	cmake.configure(source_folder="cloned_repo")
	ConanException: Error 256 while executing cd '/Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCONAN_EXPORTED="1" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="9.0" -DCONAN_CXX_FLAGS="-m64" -DCONAN_SHARED_LINKER_FLAGS="-m64" -DCONAN_C_FLAGS="-m64" -DCONAN_LIBCXX="libc++" -DCMAKE_INSTALL_PREFIX="/Users/adn/.conan/data/scm-test/0.0.0/com/develop/package/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c" -Wno-dev '/Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/cloned_repo'

So install fails complaining .../cloned_repo does not contain the CMakeLists.txt, and for a good reason:

$ ls -1a /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/cloned_repo/
.
..
$ ls -1a /Users/adn/.conan/data/scm-test/0.0.0/com/develop/build/7fe67dff831b24bc4a8b5db678a51f1be5e44e7c/
.
..
.git
.gitignore
CMakeLists.txt
build
cloned_repo
conan_paths.cmake
conanbuildinfo.cmake
main.cpp

--- a/conans/client/cmd/export.py
+++ b/conans/client/cmd/export.py
@@ -9,11 +9,11 @@ from conans.client.cmd.export_linter import conan_linter
 from conans.client.file_copier import FileCopier
 from conans.client.loader_parse import load_conanfile_class
 from conans.client.output import ScopedOutput
-from conans.client.source import get_scm
 from conans.errors import ConanException
 from conans.model.conan_file import create_exports, create_exports_sources
 from conans.model.manifest import FileTreeManifest
 from conans.model.ref import ConanFileReference
+from conans.model.scm import SCM, SCMData
 from conans.paths import CONAN_MANIFEST, CONANFILE
 from conans.util.files import save, rmdir, is_dirty, set_dirty, mkdir
 from conans.util.log import logger
@@ -104,24 +104,29 @@ def _capture_export_scm_data(conanfile, src_path, destination_folder, output, pa
     if os.path.exists(scm_src_file):
         os.unlink(scm_src_file)
 
-    scm = get_scm(conanfile, src_path)
-
-    if not scm or not (scm.capture_origin or scm.capture_revision):
+    try:
+        scm_data = SCMData(conanfile)
+    except ConanException:
         return
+    else:
+        if not (scm_data.capture_origin or scm_data.capture_revision):
+            return
 
-    if scm.url == "auto":
+    scm = SCM(scm_data, src_path)
+
+    if scm_data.url == "auto":
         origin = scm.get_remote_url()
         if not origin:
             raise ConanException("Repo origin cannot be deduced by 'auto'")
         output.success("Repo origin deduced by 'auto': %s" % origin)
-        scm.url = origin
-    if scm.revision == "auto":
-        scm.revision = scm.get_revision()
-        output.success("Revision deduced by 'auto': %s" % scm.revision)
+        scm_data.url = origin
+    if scm_data.revision == "auto":
+        scm_data.revision = scm.get_revision()
+        output.success("Revision deduced by 'auto': %s" % scm_data.revision)
 
     # Generate the scm_folder.txt file pointing to the src_path
     save(scm_src_file, src_path.replace("\\", "/"))
-    scm.replace_in_file(os.path.join(destination_folder, "conanfile.py"))
+    scm_data.replace_in_file(os.path.join(destination_folder, "conanfile.py"))
 
 
 def _export_conanfile(conanfile_path, output, paths, conanfile, conan_ref, keep_source):

--- a/conans/client/source.py
+++ b/conans/client/source.py
@@ -36,16 +36,17 @@ def complete_recipe_sources(remote_manager, client_cache, registry, conanfile, c
 
 def merge_directories(src, dst, excluded=None, symlinks=True):
     excluded = excluded or []
+    excluded = [os.path.normpath(entry) for entry in excluded]
 
     for src_dir, dirs, files in os.walk(src, followlinks=True):
 
         def is_excluded(root_folder, element):
-            rel_path = os.path.relpath(os.path.join(root_folder, element), src)
+            rel_path = os.path.normpath(os.path.relpath(os.path.join(root_folder, element), src))
             if rel_path in excluded:
                 return True
             # If dst is a sub-folder of src_dir, discard it as an src file
             abs_path = os.path.join(root_folder, element)
-            return os.path.commonprefix([abs_path, dst]) == dst
+            return os.path.normpath(os.path.commonprefix([abs_path, dst])) == os.path.normpath(dst)
 
         # Overwriting the dirs will prevents walk to get into them
         dirs[:] = [d for d in dirs if not is_excluded(src_dir, d)]

--- a/conans/client/source.py
+++ b/conans/client/source.py
@@ -36,26 +36,25 @@ def complete_recipe_sources(remote_manager, client_cache, registry, conanfile, c
 
 def merge_directories(src, dst, excluded=None, symlinks=True):
     excluded = excluded or []
-    cmn_prefix = os.path.commonprefix([src, dst])
+
     for src_dir, dirs, files in os.walk(src, followlinks=True):
-        def is_excluded(the_dir):
-            if the_dir in excluded:
+
+        def is_excluded(root_folder, element):
+            rel_path = os.path.relpath(os.path.join(root_folder, element), src)
+            if rel_path in excluded:
                 return True
-            # If dst is a sub-folder of src_dir avoid infinite loop
-            if cmn_prefix == src:  # dst is a subfolder
-                rel_dst = os.path.relpath(dst, src)
-                return rel_dst.split(os.path.sep)[0] == the_dir
+            # If dst is a sub-folder of src_dir, discard it as an src file
+            abs_path = os.path.join(root_folder, element)
+            return os.path.commonprefix([abs_path, dst]) == dst
+
         # Overwriting the dirs will prevents walk to get into them
-        if src_dir == src:  # First step, root folder
-            dirs[:] = [d for d in dirs if not is_excluded(d)]
-            files[:] = [d for d in files if not is_excluded(d)]
+        dirs[:] = [d for d in dirs if not is_excluded(src_dir, d)]
+        files[:] = [d for d in files if not is_excluded(src_dir, d)]
 
         dst_dir = os.path.normpath(os.path.join(dst, os.path.relpath(src_dir, src)))
         if not os.path.exists(dst_dir):
             os.makedirs(dst_dir)
         for file_ in files:
-            if os.path.join(src, src_dir, file_) == dst:
-                continue
             src_file = os.path.join(src_dir, file_)
             dst_file = os.path.join(dst_dir, file_)
             if os.path.islink(src_file) and symlinks:

--- a/conans/client/source.py
+++ b/conans/client/source.py
@@ -10,15 +10,7 @@ from conans.model.conan_file import get_env_context_manager
 from conans.model.scm import SCM
 from conans.paths import EXPORT_TGZ_NAME, EXPORT_SOURCES_TGZ_NAME, CONANFILE, CONAN_MANIFEST
 from conans.util.files import rmdir, set_dirty, is_dirty, clean_dirty, mkdir
-
-
-def get_scm(conanfile, src_folder):
-    data = getattr(conanfile, "scm", None)
-    if data is not None and isinstance(data, dict):
-        return SCM(data, src_folder)
-    else:
-        # not an instance of dict or None, skip SCM feature.
-        pass
+from conans.model.scm import SCMData
 
 
 def complete_recipe_sources(remote_manager, client_cache, registry, conanfile, conan_reference):
@@ -42,15 +34,35 @@ def complete_recipe_sources(remote_manager, client_cache, registry, conanfile, c
                                       current_remote)
 
 
-def merge_directories(src, dst):
-    for src_dir, _, files in os.walk(src):
+def merge_directories(src, dst, excluded=None, symlinks=True):
+    excluded = excluded or []
+    cmn_prefix = os.path.commonprefix([src, dst])
+    for src_dir, dirs, files in os.walk(src, followlinks=True):
+        def is_excluded(the_dir):
+            if the_dir in excluded:
+                return True
+            # If dst is a sub-folder of src_dir avoid infinite loop
+            if cmn_prefix == src:  # dst is a subfolder
+                rel_dst = os.path.relpath(dst, src)
+                return rel_dst.split(os.path.sep)[0] == the_dir
+        # Overwriting the dirs will prevents walk to get into them
+        if src_dir == src:  # First step, root folder
+            dirs[:] = [d for d in dirs if not is_excluded(d)]
+            files[:] = [d for d in files if not is_excluded(d)]
+
         dst_dir = os.path.join(dst, os.path.relpath(src_dir, src))
         if not os.path.exists(dst_dir):
             os.makedirs(dst_dir)
         for file_ in files:
+            if os.path.join(src, src_dir, file_) == dst:
+                continue
             src_file = os.path.join(src_dir, file_)
             dst_file = os.path.join(dst_dir, file_)
-            shutil.copy2(src_file, dst_file)
+            if os.path.islink(src_file) and symlinks:
+                linkto = os.readlink(src_file)
+                os.symlink(linkto, dst_file)
+            else:
+                shutil.copy2(src_file, dst_file)
 
 
 def _clean_source_folder(folder):
@@ -63,7 +75,7 @@ def _clean_source_folder(folder):
 
 
 def config_source(export_folder, export_source_folder, local_sources_path, src_folder,
-                  conan_file, output, force=False):
+                  conanfile, output, force=False):
     """ creates src folder and retrieve, calling source() from conanfile
     the necessary source code
     """
@@ -88,43 +100,43 @@ def config_source(export_folder, export_source_folder, local_sources_path, src_f
     elif is_dirty(src_folder):
         output.warn("Trying to remove corrupted source folder")
         remove_source()
-    elif conan_file.build_policy_always:
+    elif conanfile.build_policy_always:
         output.warn("Detected build_policy 'always', trying to remove source folder")
         remove_source()
 
     if not os.path.exists(src_folder):
         output.info('Configuring sources in %s' % src_folder)
-        shutil.copytree(export_folder, src_folder, symlinks=True)
-        # Now move the export-sources to the right location
-        merge_directories(export_source_folder, src_folder)
-        _clean_source_folder(src_folder)
-        try:
-            shutil.rmtree(os.path.join(src_folder, "__pycache__"))
-        except OSError:
-            pass
-
         set_dirty(src_folder)
+        mkdir(src_folder)
         os.chdir(src_folder)
-        conan_file.source_folder = src_folder
+        conanfile.source_folder = src_folder
         try:
-            with get_env_context_manager(conan_file):
-                with conanfile_exception_formatter(str(conan_file), "source"):
-                    conan_file.build_folder = None
-                    conan_file.package_folder = None
+            with conanfile_exception_formatter(str(conanfile), "source"):
+                with get_env_context_manager(conanfile):
+                    conanfile.build_folder = None
+                    conanfile.package_folder = None
 
-                    scm = get_scm(conan_file, src_folder)
-                    if scm:
-                        # scm.capture_origin before exporting
-                        if local_sources_path and os.path.exists(local_sources_path):
-                            output.info("Getting sources from folder: %s" % local_sources_path)
-                            merge_directories(local_sources_path, src_folder)
-                            _clean_source_folder(src_folder)
-                        else:
-                            output.info("Getting sources from url: '%s'" % scm.url)
-                            scm.clone()
-                            scm.checkout()
+                    try:
+                        scm_data = SCMData(conanfile)
+                    except ConanException:
+                        pass
+                    else:
+                        dest_dir = os.path.join(src_folder, scm_data.subfolder)
+                        captured = local_sources_path and os.path.exists(local_sources_path)
+                        local_sources_path = local_sources_path if captured else None
+                        _fetch_scm(scm_data, dest_dir, local_sources_path, output)
 
-                    conan_file.source()
+                    merge_directories(export_folder, src_folder)
+                    # Now move the export-sources to the right location
+                    merge_directories(export_source_folder, src_folder)
+                    _clean_source_folder(src_folder)
+                    try:
+                        shutil.rmtree(os.path.join(src_folder, "__pycache__"))
+                    except OSError:
+                        pass
+
+                    conanfile.source()
+
             clean_dirty(src_folder)  # Everything went well, remove DIRTY flag
         except Exception as e:
             os.chdir(export_folder)
@@ -137,28 +149,41 @@ def config_source(export_folder, export_source_folder, local_sources_path, src_f
             raise ConanException(e)
 
 
-def config_source_local(dest_dir, conan_file, conanfile_folder, output):
+def config_source_local(dest_dir, conanfile, conanfile_folder, output):
     output.info('Configuring sources in %s' % dest_dir)
-    conan_file.source_folder = dest_dir
+    conanfile.source_folder = dest_dir
     with tools.chdir(dest_dir):
         try:
-            with conanfile_exception_formatter(str(conan_file), "source"):
-                with get_env_context_manager(conan_file):
-                    conan_file.build_folder = None
-                    conan_file.package_folder = None
-                    scm = get_scm(conan_file, dest_dir)
-                    if scm:
-                        if scm.capture_origin or scm.capture_revision:
-                            output.info("Getting sources from folder: %s" % conanfile_folder)
-                            merge_directories(conanfile_folder, dest_dir)
-                            _clean_source_folder(dest_dir)
-                        else:
-                            output.info("Getting sources from url: '%s'" % scm.url)
-                            scm.clone()
-                            scm.checkout()
+            with conanfile_exception_formatter(str(conanfile), "source"):
+                with get_env_context_manager(conanfile):
+                    conanfile.build_folder = None
+                    conanfile.package_folder = None
+                    try:
+                        scm_data = SCMData(conanfile)
+                    except ConanException:
+                        pass
+                    else:
+                        dest_dir = os.path.join(dest_dir, scm_data.subfolder)
+                        capture = scm_data.capture_origin or scm_data.capture_revision
+                        local_sources_path = conanfile_folder if capture else None
+                        _fetch_scm(scm_data, dest_dir, local_sources_path, output)
 
-                    conan_file.source()
+                    conanfile.source()
         except ConanExceptionInUserConanfileMethod:
             raise
         except Exception as e:
             raise ConanException(e)
+
+
+def _fetch_scm(scm_data, dest_dir, local_sources_path, output):
+    if local_sources_path:
+        excluded = SCM(scm_data, local_sources_path).excluded_files
+        output.info("Getting sources from folder: %s" % local_sources_path)
+        output.info("Excluding files: %s" % ", ".join(excluded))
+        merge_directories(local_sources_path, dest_dir, excluded=excluded)
+    else:
+        output.info("Getting sources from url: '%s'" % scm_data.url)
+        scm = SCM(scm_data, dest_dir)
+        scm.clone()
+        scm.checkout()
+    _clean_source_folder(dest_dir)

--- a/conans/client/source.py
+++ b/conans/client/source.py
@@ -50,7 +50,7 @@ def merge_directories(src, dst, excluded=None, symlinks=True):
             dirs[:] = [d for d in dirs if not is_excluded(d)]
             files[:] = [d for d in files if not is_excluded(d)]
 
-        dst_dir = os.path.join(dst, os.path.relpath(src_dir, src))
+        dst_dir = os.path.normpath(os.path.join(dst, os.path.relpath(src_dir, src)))
         if not os.path.exists(dst_dir):
             os.makedirs(dst_dir)
         for file_ in files:
@@ -121,7 +121,7 @@ def config_source(export_folder, export_source_folder, local_sources_path, src_f
                     except ConanException:
                         pass
                     else:
-                        dest_dir = os.path.join(src_folder, scm_data.subfolder)
+                        dest_dir = os.path.normpath(os.path.join(src_folder, scm_data.subfolder))
                         captured = local_sources_path and os.path.exists(local_sources_path)
                         local_sources_path = local_sources_path if captured else None
                         _fetch_scm(scm_data, dest_dir, local_sources_path, output)

--- a/conans/client/tools/scm.py
+++ b/conans/client/tools/scm.py
@@ -1,5 +1,6 @@
 import os
 import subprocess
+from subprocess import CalledProcessError
 
 from six.moves.urllib.parse import urlparse, quote_plus
 
@@ -50,10 +51,10 @@ class Git(object):
             url = url.replace("\\", "/")  # Windows local directory
         if os.path.exists(self.folder) and os.listdir(self.folder):
             if not branch:
-                raise ConanException("The destination folder is not empty, "
+                raise ConanException("The destination folder '%s' is not empty, "
                                      "specify a branch to checkout (not a tag or commit) "
-                                     "or specify a 'subfolder'"
-                                     "attribute in the 'scm'")
+                                     "or specify a 'subfolder' "
+                                     "attribute in the 'scm'" % self.folder)
             output = self.run("init")
             output += self._configure_ssl_verify()
             output += self.run('remote add origin "%s"' % url)
@@ -68,6 +69,14 @@ class Git(object):
     def checkout(self, element):
         # Element can be a tag, branch or commit
         return self.run('checkout "%s"' % element)
+
+    def excluded_files(self):
+        try:
+            tmp = self.run("check-ignore *").splitlines()
+        except CalledProcessError:
+            tmp = []
+        tmp.append(".git")
+        return tmp
 
     def get_remote_url(self, remote_name=None):
         remote_name = remote_name or "origin"

--- a/conans/client/tools/scm.py
+++ b/conans/client/tools/scm.py
@@ -7,7 +7,7 @@ from six.moves.urllib.parse import urlparse, quote_plus
 from conans.client.tools.env import no_op, environment_append
 from conans.client.tools.files import chdir
 from conans.errors import ConanException
-from conans.util.files import decode_text
+from conans.util.files import decode_text, to_file_bytes
 
 
 class Git(object):
@@ -77,7 +77,8 @@ class Git(object):
                           for folder, _, fs in os.walk(self.folder) for f in fs]
             p = subprocess.Popen(['git', 'check-ignore', '--stdin'],
                                  stdout=PIPE, stdin=PIPE, stderr=STDOUT, cwd=self.folder)
-            grep_stdout = decode_text(p.communicate(input='\n'.join(file_paths))[0])
+            paths = to_file_bytes('\n'.join(file_paths))
+            grep_stdout = decode_text(p.communicate(input=paths)[0])
             tmp = grep_stdout.splitlines()
         except CalledProcessError:
             tmp = []

--- a/conans/client/tools/scm.py
+++ b/conans/client/tools/scm.py
@@ -1,12 +1,13 @@
 import os
 import subprocess
-from subprocess import CalledProcessError
+from subprocess import CalledProcessError, PIPE, STDOUT
 
 from six.moves.urllib.parse import urlparse, quote_plus
 
 from conans.client.tools.env import no_op, environment_append
 from conans.client.tools.files import chdir
 from conans.errors import ConanException
+from conans.util.files import decode_text
 
 
 class Git(object):
@@ -72,7 +73,12 @@ class Git(object):
 
     def excluded_files(self):
         try:
-            tmp = self.run("check-ignore *").splitlines()
+            file_paths = [os.path.normpath(os.path.join(os.path.relpath(folder, self.folder), f))
+                          for folder, _, fs in os.walk(self.folder) for f in fs]
+            p = subprocess.Popen(['git', 'check-ignore', '--stdin'],
+                                 stdout=PIPE, stdin=PIPE, stderr=STDOUT, cwd=self.folder)
+            grep_stdout = decode_text(p.communicate(input='\n'.join(file_paths))[0])
+            tmp = grep_stdout.splitlines()
         except CalledProcessError:
             tmp = []
         tmp.append(".git")

--- a/conans/model/scm.py
+++ b/conans/model/scm.py
@@ -1,5 +1,4 @@
 import json
-import os
 
 from conans.client.tools.scm import Git
 from conans.errors import ConanException
@@ -7,31 +6,20 @@ from conans.util.files import load, save
 import re
 
 
-class SCM(object):
+class SCMData(object):
 
-    def __init__(self, data, src_folder):
-        self.type = data.get("type")
-        self.url = data.get("url")
-        self.revision = data.get("revision")
-        self.verify_ssl = data.get("verify_ssl")
-        self.subfolder = data.get("subfolder")
-        if data.get("subfolder"):
-            self.src_folder = os.path.join(src_folder, self.subfolder)
+    def __init__(self, conanfile):
+        data = getattr(conanfile, "scm", None)
+        if data is not None and isinstance(data, dict):
+            self.type = data.get("type")
+            self.url = data.get("url")
+            self.revision = data.get("revision")
+            self.verify_ssl = data.get("verify_ssl")
+            self.username = data.get("username")
+            self.password = data.get("password")
+            self.subfolder = data.get("subfolder", "")
         else:
-            self.src_folder = src_folder
-
-        self.username = data.get("username")
-        self.password = data.get("password")
-
-        # Finally instance a repo
-        self.repo = self._get_repo()
-
-    def _get_repo(self):
-        repo = {"git": Git(self.src_folder, verify_ssl=self.verify_ssl,
-                           username=self.username, password=self.password)}.get(self.type)
-        if not repo:
-            raise ConanException("SCM not supported: %s" % self.type)
-        return repo
+            raise ConanException("Not SCM enabled in conanfile")
 
     @property
     def capture_origin(self):
@@ -40,18 +28,6 @@ class SCM(object):
     @property
     def capture_revision(self):
         return self.revision == "auto"
-
-    def clone(self):
-        return self.repo.clone(self.url)
-
-    def checkout(self):
-        return self.repo.checkout(self.revision)
-
-    def get_remote_url(self):
-        return self.repo.get_remote_url()
-
-    def get_revision(self):
-        return self.repo.get_revision()
 
     def __repr__(self):
         d = {"url": self.url, "revision": self.revision, "username": self.username,
@@ -66,3 +42,35 @@ class SCM(object):
         dumps = ",\n          ".join(dumps.split(","))
         content = re.sub(r"scm\s*=\s*{[^}]*}", "scm = %s" % dumps, content)
         save(path, content)
+
+
+class SCM(object):
+    def __init__(self, data, repo_folder):
+        self._data = data
+        self.repo_folder = repo_folder
+        # Finally instance a repo
+        self.repo = self._get_repo()
+
+    def _get_repo(self):
+        repo = {"git": Git(self.repo_folder, verify_ssl=self._data.verify_ssl,
+                           username=self._data.username,
+                           password=self._data.password)}.get(self._data.type)
+        if not repo:
+            raise ConanException("SCM not supported: %s" % self._data.type)
+        return repo
+
+    @property
+    def excluded_files(self):
+        return self.repo.excluded_files()
+
+    def clone(self):
+        return self.repo.clone(self._data.url)
+
+    def checkout(self):
+        return self.repo.checkout(self._data.revision)
+
+    def get_remote_url(self):
+        return self.repo.get_remote_url()
+
+    def get_revision(self):
+        return self.repo.get_revision()

--- a/conans/test/functional/merge_directories_test.py
+++ b/conans/test/functional/merge_directories_test.py
@@ -34,13 +34,17 @@ class MergeDirectoriesTest(unittest.TestCase):
                     ret.append(relpath)
         return ret
 
+    def _assert_equals(self, list1, list2):
+        self.assertEquals(set([el.replace("/", "\\") for el in list1]),
+                          set([el.replace("/", "\\") for el in list2]))
+
     def test_empty_dest_merge(self):
 
         files = ["file.txt", "subdir/file2.txt"]
         self._save(self.source, files)
 
         merge_directories(self.source, self.dest)
-        self.assertEquals(set(self._get_paths(self.dest)), set(files))
+        self._assert_equals(self._get_paths(self.dest), files)
 
     def test_non_empty_dest_merge(self):
         files = ["file.txt", "subdir/file2.txt"]
@@ -50,7 +54,7 @@ class MergeDirectoriesTest(unittest.TestCase):
         self._save(self.dest, files_dest, "fromdest")
 
         merge_directories(self.source, self.dest)
-        self.assertEquals(set(self._get_paths(self.dest)), set(files + files_dest))
+        self._assert_equals(self._get_paths(self.dest), files + files_dest)
         # File from src overrides file from dest
         self.assertEquals(load(join(self.dest, "file.txt")), "fromsrc")
 
@@ -64,9 +68,8 @@ class MergeDirectoriesTest(unittest.TestCase):
         self._save(self.source, files, "fromsrc")
 
         merge_directories(self.source, self.dest)
-        self.assertEquals(set(self._get_paths(self.dest)), set(files +
-                                                               files_dest +
-                                                               ['empty_folder/subempty_folder', ]))
+        self._assert_equals(self._get_paths(self.dest), files + files_dest +
+                           ['empty_folder/subempty_folder', ])
 
     def excluded_dirs_test(self):
         files = ["file.txt", "subdir/file2.txt", "subdir/file3.txt", "other_dir/somefile.txt",
@@ -78,15 +81,15 @@ class MergeDirectoriesTest(unittest.TestCase):
 
         # Excluded one file from other_dir and the whole subdir
         merge_directories(self.source, self.dest, excluded=["other_dir/somefile.txt", "subdir"])
-        self.assertEquals(set(self._get_paths(self.dest)), set(["file.txt",
-                                                                "subdir2/file2.txt",
-                                                                "other_dir/somefile2.txt"]))
+        self._assert_equals(self._get_paths(self.dest), ["file.txt",
+                                                         "subdir2/file2.txt",
+                                                         "other_dir/somefile2.txt"])
 
         # Excluded one from dest (no sense) and one from origin
         merge_directories(self.source, self.dest, excluded=["subdir2/file2.txt",
                                                             "subdir",
                                                             "other_dir/somefile.txt"])
 
-        self.assertEquals(set(self._get_paths(self.dest)), set(["file.txt",
-                                                                "subdir2/file2.txt",
-                                                                "other_dir/somefile2.txt"]))
+        self._assert_equals(self._get_paths(self.dest), ["file.txt",
+                                                         "subdir2/file2.txt",
+                                                         "other_dir/somefile2.txt"])

--- a/conans/test/functional/merge_directories_test.py
+++ b/conans/test/functional/merge_directories_test.py
@@ -1,0 +1,92 @@
+# coding=utf-8
+import unittest
+import os
+
+from os.path import join
+
+from conans import load
+from conans.client.source import merge_directories
+from conans.util.files import save, mkdir
+from conans.test.utils.test_files import temp_folder
+
+
+class MergeDirectoriesTest(unittest.TestCase):
+
+    def setUp(self):
+        self.source = temp_folder()
+        self.dest = temp_folder()
+
+    def _save(self, folder, files, content=None):
+        for the_file in files:
+            save(join(folder, the_file), content or "c")
+
+    def _get_paths(self, folder):
+        ret = []
+        for root, dirs, files in os.walk(folder):
+            for the_file in files:
+                relpath = os.path.relpath(root, folder)
+                if relpath == ".":
+                    relpath = ""
+                ret.append(join(relpath, the_file))
+            for the_dir in dirs:
+                if not os.listdir(join(root, the_dir)):
+                    relpath = os.path.relpath(join(root, the_dir), folder)
+                    ret.append(relpath)
+        return ret
+
+    def test_empty_dest_merge(self):
+
+        files = ["file.txt", "subdir/file2.txt"]
+        self._save(self.source, files)
+
+        merge_directories(self.source, self.dest)
+        self.assertEquals(set(self._get_paths(self.dest)), set(files))
+
+    def test_non_empty_dest_merge(self):
+        files = ["file.txt", "subdir/file2.txt"]
+        self._save(self.source, files, "fromsrc")
+
+        files_dest = ["file.txt", "subdir2/file2.txt"]
+        self._save(self.dest, files_dest, "fromdest")
+
+        merge_directories(self.source, self.dest)
+        self.assertEquals(set(self._get_paths(self.dest)), set(files + files_dest))
+        # File from src overrides file from dest
+        self.assertEquals(load(join(self.dest, "file.txt")), "fromsrc")
+
+    def nested_directories_test(self):
+        self.dest = join(self.source, "destination_dir")
+        files_dest = ["file.txt", "subdir2/file2.txt"]
+        self._save(self.dest, files_dest, "fromdest")
+        mkdir(join(self.dest, "empty_folder", "subempty_folder"))
+
+        files = ["file.txt", "subdir/file2.txt"]
+        self._save(self.source, files, "fromsrc")
+
+        merge_directories(self.source, self.dest)
+        self.assertEquals(set(self._get_paths(self.dest)), set(files +
+                                                               files_dest +
+                                                               ['empty_folder/subempty_folder', ]))
+
+    def excluded_dirs_test(self):
+        files = ["file.txt", "subdir/file2.txt", "subdir/file3.txt", "other_dir/somefile.txt",
+                 "other_dir/somefile2.txt"]
+        self._save(self.source, files, "fromsrc")
+
+        files_dest = ["file.txt", "subdir2/file2.txt"]
+        self._save(self.dest, files_dest, "fromdest")
+
+        # Excluded one file from other_dir and the whole subdir
+        merge_directories(self.source, self.dest, excluded=["other_dir/somefile.txt", "subdir"])
+        self.assertEquals(set(self._get_paths(self.dest)), set(["file.txt",
+                                                                "subdir2/file2.txt",
+                                                                "other_dir/somefile2.txt"]))
+
+        # Excluded one from dest (no sense) and one from origin
+        merge_directories(self.source, self.dest, excluded=["subdir2/file2.txt",
+                                                            "subdir",
+                                                            "other_dir/somefile.txt"])
+
+        self.assertEquals(set(self._get_paths(self.dest)), set(["file.txt",
+                                                                "subdir2/file2.txt",
+                                                                "other_dir/somefile2.txt"]))

--- a/conans/test/functional/scm_test.py
+++ b/conans/test/functional/scm_test.py
@@ -121,6 +121,7 @@ class ConanLib(ConanFile):
         conanfile = base.replace('"revision": "{revision}"',
                                  '"revision": "{revision}",\n        '
                                  '"subfolder": "mysub"')
+        conanfile = conanfile.replace("short_paths = True", "short_paths = False")
         conanfile = conanfile.format(directory="None", url="auto", revision="auto")
         self.client.save({"conanfile.py": conanfile, "myfile.txt": "My file is copied"})
         self._commit_contents()
@@ -150,6 +151,7 @@ class ConanLib(ConanFile):
 
     def test_excluded_repo_fies(self):
         conanfile = base.format(url="auto", revision="auto")
+        conanfile = conanfile.replace("short_paths = True", "short_paths = False")
         path, commit = create_local_git_repo({"myfile": "contents",
                                               "ignored.pyc": "bin",
                                               ".gitignore": "*.pyc\n",

--- a/conans/test/functional/scm_test.py
+++ b/conans/test/functional/scm_test.py
@@ -1,15 +1,17 @@
 import json
 import os
 import unittest
+from collections import namedtuple
 
 from conans.client.tools.scm import Git
-from conans.model.ref import ConanFileReference
-from conans.model.scm import SCM
+from conans.model.ref import ConanFileReference, PackageReference
+from conans.model.scm import SCM, SCMData
 from conans.test.utils.test_files import temp_folder
 from conans.test.utils.tools import TestClient, TestServer, create_local_git_repo
 from conans.util.files import load, rmdir
 
 base = '''
+import os
 from conans import ConanFile, tools
 
 class ConanLib(ConanFile):
@@ -23,7 +25,9 @@ class ConanLib(ConanFile):
     }}
 
     def build(self):
-        self.output.warn(tools.load("myfile.txt"))
+        sf = self.scm.get("subfolder")
+        path = os.path.join(sf, "myfile.txt") if sf else "myfile.txt"
+        self.output.warn(tools.load(path))
 '''
 
 
@@ -78,6 +82,7 @@ class ConanLib(ConanFile):
         self.client.run("create . user/channel")
         folder = self.client.client_cache.source(ConanFileReference.loads("lib/0.1@user/channel"))
         self.assertIn("othersubfolder", os.listdir(folder))
+        self.assertTrue(os.path.exists(os.path.join(folder, "othersubfolder", "myfile")))
 
     def test_auto_git(self):
         curdir = self.client.current_folder.replace("\\", "/")
@@ -111,6 +116,21 @@ class ConanLib(ConanFile):
         self.assertIn("Getting sources from url: '%s'" % curdir, self.client.out)
         self.assertIn("My file is copied", self.client.out)
 
+    def test_auto_subfolder(self):
+        curdir = self.client.current_folder.replace("\\", "/")
+        conanfile = base.replace('"revision": "{revision}"',
+                                 '"revision": "{revision}",\n        '
+                                 '"subfolder": "mysub"')
+        conanfile = conanfile.format(directory="None", url="auto", revision="auto")
+        self.client.save({"conanfile.py": conanfile, "myfile.txt": "My file is copied"})
+        self._commit_contents()
+        self.client.runner('git remote add origin https://myrepo.com.git', cwd=curdir)
+        self.client.run("create . user/channel")
+
+        folder = self.client.client_cache.source(ConanFileReference.loads("lib/0.1@user/channel"))
+        self.assertTrue(os.path.exists(os.path.join(folder, "mysub", "myfile.txt")))
+        self.assertFalse(os.path.exists(os.path.join(folder, "mysub", "conanfile.py")))
+
     def test_deleted_source_folder(self):
         path, commit = create_local_git_repo({"myfile": "contents"}, branch="my_release")
         curdir = self.client.current_folder.replace("\\", "/")
@@ -127,6 +147,27 @@ class ConanLib(ConanFile):
         error = self.client.run("install lib/0.1@user/channel --build", ignore_error=True)
         self.assertTrue(error)
         self.assertIn("Getting sources from url: '%s'" % path.replace("\\", "/"), self.client.out)
+
+    def test_excluded_repo_fies(self):
+        conanfile = base.format(url="auto", revision="auto")
+        path, commit = create_local_git_repo({"myfile": "contents",
+                                              "ignored.pyc": "bin",
+                                              ".gitignore": "*.pyc\n",
+                                              "myfile.txt": "My file!",
+                                              "conanfile.py": conanfile}, branch="my_release")
+        self.client.current_folder = path
+        self._commit_contents()
+        self.client.runner('git remote add origin "%s"' % path.replace("\\", "/"), cwd=path)
+        self.client.run("create . user/channel")
+        self.assertIn("Excluding files: ignored.pyc, .git", self.client.out)
+        self.assertIn("Copying sources to build folder", self.client.out)
+        pref = PackageReference(ConanFileReference.loads("lib/0.1/user/channel"),
+                                "5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9")
+        bf = self.client.client_cache.build(pref)
+        self.assertTrue(os.path.exists(os.path.join(bf, "myfile.txt")))
+        self.assertTrue(os.path.exists(os.path.join(bf, "myfile")))
+        self.assertFalse(os.path.exists(os.path.join(bf, ".git")))
+        self.assertFalse(os.path.exists(os.path.join(bf, "ignored.pyc")))
 
     def test_local_source(self):
         curdir = self.client.current_folder
@@ -161,6 +202,24 @@ class ConanLib(ConanFile):
         self.assertFalse(os.path.exists(os.path.join(curdir, "source2", "myfile2.txt")))
         self.assertTrue(os.path.exists(os.path.join(curdir, "source2", "myfile.txt")))
         self.assertIn("Getting sources from url: '%s'" % curdir.replace("\\", "/"), self.client.out)
+        self.assertIn("SOURCE METHOD CALLED", self.client.out)
+
+    def test_local_source_subfolder(self):
+        curdir = self.client.current_folder
+        conanfile = base.replace('"revision": "{revision}"',
+                                 '"revision": "{revision}",\n        '
+                                 '"subfolder": "mysub"')
+        conanfile = conanfile.format(url="auto", revision="auto")
+        conanfile += """
+    def source(self):
+        self.output.warn("SOURCE METHOD CALLED")
+"""
+        self.client.save({"conanfile.py": conanfile, "myfile.txt": "My file is copied"})
+        self._commit_contents()
+
+        self.client.run("source . --source-folder=./source")
+        self.assertFalse(os.path.exists(os.path.join(curdir, "source", "myfile.txt")))
+        self.assertTrue(os.path.exists(os.path.join(curdir, "source", "mysub", "myfile.txt")))
         self.assertIn("SOURCE METHOD CALLED", self.client.out)
 
     def test_install_checked_out(self):
@@ -220,7 +279,8 @@ class ConanLib(ConanFile):
         data = {"url": "myurl", "revision": "myrevision", "username": "myusername",
                 "password": "mypassword", "type": "git", "verify_ssl": True,
                 "subfolder": "mysubfolder"}
-        scm = SCM(data, temp_folder())
-        the_json = str(scm)
+        conanfile = namedtuple("ConanfileMock", "scm")(data)
+        scm_data = SCMData(conanfile)
+        the_json = str(scm_data)
         data2 = json.loads(the_json)
         self.assertEquals(data, data2)

--- a/conans/test/util/tools_test.py
+++ b/conans/test/util/tools_test.py
@@ -929,8 +929,7 @@ class GitToolTest(unittest.TestCase):
         tmp = temp_folder()
         save(os.path.join(tmp, "file"), "dummy contents")
         git = Git(tmp)
-        with self.assertRaisesRegexp(ConanException, "The destination folder is not empty, "
-                                                     "specify a branch to checkout"):
+        with self.assertRaisesRegexp(ConanException, "specify a branch to checkout"):
             git.clone("https://github.com/conan-community/conan-zlib.git")
 
     def test_credentials(self):
@@ -1032,5 +1031,4 @@ class HelloConan(ConanFile):
 """ % git_repo.replace("\\", "/")
         client.save({"conanfile.py": conanfile, "other": "hello"})
         client.run("create . user/channel", ignore_error=True)
-        self.assertIn("The destination folder is not empty, "
-                      "specify a branch to checkout", client.out)
+        self.assertIn("specify a branch to checkout", client.out)

